library/src/main/scala/scala/scalajs/runtime/RuntimeLong.scala

```diff
@@ -302,9 +302,212 @@ final class RuntimeLong(val lo: Int, val hi: Int)
+     *   = a1b1 + (2^16*a1b0 + 2^16*a0b1 + a0b0) / 2^32
+     *   = a1b1 + (2^16*a1b0 + 2^16*a0b1 + (2^16*(a0b0 / 2^16) + (a0b0 % 2^16))) / 2^32
+     *   = a1b1 + (2^16*(a1b0 + a0b1 + (a0b0 / 2^16)) + (a0 % 2^16)) / 2^32
+     *             ----------------------------------
+     *                  multiple of 2^16
+     *   = a1b1 + ( (2^16*(a1b0 + a0b1 + (a0b0 / 2^16))) / 2^16 + (a0 % 2^16) / 2^16 ) / 2^16
+     *                                                             ---------
+     *                                                               < 2^16
+     *   = a1b1 + (a1b0 + a0b1 + (a0b0 / 2^16)) / 2^16
+     *   = a1b1 + (a1b0 + (a0b1 + (a0b0 >>> 16))) / 2^16
+     *                     ----    -----------
+     *                <= fffe0001    <= ffff
+     *                     --------------------
+     *                         <= ffff0000, hence the + does not overflow
+     *   = a1b1 + (a1b0 + (a0b1 +[32] (a0b0 >>> 16))) / 2^16
+     *
+     * Let
+     *   c1part = a0b1 +[32] (a0b0 >>> 16)
+     *
+     * carry_from_lo_*
+     *   = a1b1 + (a1b0 + c1part) / 2^16
+     *   = a1b1 + (a1b0 + (2^16*(c1part / 2^16) + (c1part % 2^16))) / 2^16
+     *   = a1b1 + (2^16*(c1part / 2^16) + (a1b0 + (c1part % 2^16))) / 2^16
+     *   = a1b1 + (2^16*(c1part / 2^16) + (a1b0 + (c1part & 0xffff))) / 2^16
+     *                                     ----    ---------------
+     *                                <= fffe0001     <= ffff
+     *                                     ---------------------
+     *                                <= ffff0000, hence the + does not overflow
+     *   = a1b1 + (2^16*(c1part / 2^16) + (a1b0 +[32] (c1part & 0xffff))) / 2^16
+     *             --------------------
+     *               multiple of 1^16
+     *   = a1b1 + ( 2^16*(c1part / 2^16) / 2^16 + (a1b0 +[32] (c1part & 0xffff)) / 2^16 )
+     *   = a1b1 + (c1part / 2^16) + (a1b0 +[32] (c1part & 0xffff)) / 2^16
+     *             ------           ----------------------------
+     *             < 2^32                      < 2^32
+     *   = a1b1 + (c1part >>> 16) + ((a1b0 +[32] (c1part & 0xffff)) >>> 16)
+     *
+     * Recap so far:
+     *
+     * a *[64] b
+     *   = ( 2^32*(ahi *[32] blo) + 2^32*(alo *[32] bhi) + alo*blo ) % 2^64
+     * alo*blo
+     *   = 2^32*carry_from_lo_* + aloblo
+     * aloblo
+     *   = (a1b0 << 16) +[32] (a0b1 << 16) +[32] a0b0
+     * carry_from_lo_*
+     *   = a1b1 + (c1part >>> 16) + ((a1b0 +[32] (c1part & 0xffff)) >>> 16)
+     *
+     * Substituting,
+     *
+     * a *[64] b
+     *   = ( 2^32*(ahi *[32] blo) + 2^32*(alo *[32] bhi) + 2^32*carry_from_lo_* + aloblo ) % 2^64
+     *   = ( 2^32*((ahi *[32] blo) + (alo *[32] bhi) + carry_from_lo_*) + aloblo ) % 2^64
+     *   = ( 2^32*((ahi *[32] blo) + (alo *[32] bhi) + carry_from_lo_*) % 2^64 + aloblo ) % 2^64
+     *   = ( 2^32*(((ahi *[32] blo) + (alo *[32] bhi) + carry_from_lo_*) % 2^32) + aloblo ) % 2^64
```

Comment:
This step (`n*x % n^2 = n * (x % n)`) is non-obvious, can you make this more explicit?